From 3526721b45bd53556b046fcd19e39d7c58af4631 Mon Sep 17 00:00:00 2001
From: Jakub Jirutka <jakub@jirutka.cz>
Date: Sat, 20 May 2017 21:44:42 +0200
Subject: [PATCH] Don't fetch grpc and google-protobuf from RubyGems

These two gems from Google have platform-specific variants on RubyGems,
built against glibc. Rubygems prefers platform-specific gems when
available and I don't know how to force Bundler to pass --platform
option to gem install (to prefer platform-agnostic gems).

This will not be problem on Alpine v3.6, because I've patched Rubygems to
not fetch platform-specific gems by default.

However, there's another problem with grpc gem. It's total shit, doesn't
build from source without patching, embeds three third-party libs
including Google's very own fork of OpenSSL and root CAs, contains a lot
of bloat etc.

Therefore I install these two pieces of shit from my Alpine packages and
copy them into vendor/gems/ before running bundler.
---
 Gemfile      |  4 ++++
 Gemfile.lock | 17 +++++++++++++----
 2 files changed, 17 insertions(+), 4 deletions(-)

diff --git a/Gemfile b/Gemfile
index ee88704..5743ecc 100644
--- a/Gemfile
+++ b/Gemfile
@@ -360,4 +360,8 @@ gem 'sys-filesystem', '~> 1.1.6', group: :exclude
 # Gitaly GRPC client
 gem 'gitaly', '~> 0.5.0'
 
+# These Google gems are totally messed, we must handle them specially.
+gem 'grpc', path: 'vendor/gems/grpc', require: false
+gem 'google-protobuf', path: 'vendor/gems/google-protobuf', require: false
+
 gem 'toml-rb', '~> 0.3.15', require: false
diff --git a/Gemfile.lock b/Gemfile.lock
index 4cbfab0..60e1b11 100644
--- a/Gemfile.lock
+++ b/Gemfile.lock
@@ -1,3 +1,14 @@
+PATH
+  remote: vendor/gems/google-protobuf
+  specs:
+    google-protobuf (3.3.1)
+
+PATH
+  remote: vendor/gems/grpc
+  specs:
+    grpc (1.3.2)
+      google-protobuf (~> 3.1)
+
 GEM
   remote: https://rubygems.org/
   specs:
@@ -308,7 +319,6 @@ GEM
       multi_json (~> 1.10)
       retriable (~> 1.4)
       signet (~> 0.6)
-    google-protobuf (3.2.0.2)
     googleauth (0.5.1)
       faraday (~> 0.9)
       jwt (~> 1.4)
@@ -330,9 +340,6 @@ GEM
     grape-entity (0.6.0)
       activesupport
       multi_json (>= 1.3.2)
-    grpc (1.1.2)
-      google-protobuf (~> 3.1)
-      googleauth (~> 0.5.1)
     haml (4.0.7)
       tilt
     haml_lint (0.21.0)
@@ -912,8 +919,10 @@ DEPENDENCIES
   gollum-rugged_adapter (~> 0.4.4)
   gon (~> 6.1.0)
   google-api-client (~> 0.8.6)
+  google-protobuf!
   grape (~> 0.19.0)
   grape-entity (~> 0.6.0)
+  grpc!
   haml_lint (~> 0.21.0)
   hamlit (~> 2.6.1)
   health_check (~> 2.6.0)
-- 
2.10.1 (Apple Git-78)

